crypto/cache.go

```diff
@@ -73,127 +65,134 @@ func (cache *cache) evict() {
 
-// VerifyThresholdSignature verifies a threshold signature.
-func (cache *cache) VerifyAggregateSignature(signature consensus.ThresholdSignature, hash consensus.Hash) bool {
-	if signature == nil {
-		return false
-	}
-	key := sha256.Sum256(append(hash[:], signature.ToBytes()...))
-	if cache.check(key, false) {
-		return true
-	}
-	if cache.impl.VerifyAggregateSignature(signature, hash) {
-		cache.insert(key, false)
+	if cache.check(key.String()) {
 		return true
 	}
-	return false
-}
 
-// CreateThresholdSignature creates a threshold signature from the given partial signatures.
-func (cache *cache) CreateThresholdSignature(partialSignatures []consensus.Signature, hash consensus.Hash) (sig consensus.ThresholdSignature, err error) {
-	sig, err = cache.impl.CreateThresholdSignature(partialSignatures, hash)
-	if err != nil {
-		return nil, err
-	}
-	key := sha256.Sum256(append(hash[:], sig.ToBytes()...))
-	cache.insert(key, true)
-	return sig, nil
-}
-
-// VerifyThresholdSignature verifies a threshold signature.
-func (cache *cache) VerifyThresholdSignature(signature consensus.ThresholdSignature, hash consensus.Hash) bool {
-	if signature == nil {
-		return false
-	}
-	key := sha256.Sum256(append(hash[:], signature.ToBytes()...))
-	if cache.check(key, true) {
+	if cache.impl.Verify(signature, message) {
+		cache.insert(key.String())
 		return true
 	}
-	if cache.impl.VerifyThresholdSignature(signature, hash) {
-		cache.insert(key, true)
-		return true
-	}
-	return false
-}
 
-// CreateThresholdSignatureForMessageSet creates a threshold signature where each partial signature has signed a
-// different message hash.
-func (cache *cache) CreateThresholdSignatureForMessageSet(partialSignatures []consensus.Signature, hashes map[hotstuff.ID]consensus.Hash) (consensus.ThresholdSignature, error) {
-	signature, err := cache.impl.CreateThresholdSignatureForMessageSet(partialSignatures, hashes)
-	if err != nil {
-		return nil, err
-	}
-	var key consensus.Hash
-	hash := sha256.New()
-	for _, h := range hashes {
-		hash.Write(h[:])
-	}
-	hash.Write(signature.ToBytes())
-	hash.Sum(key[:0])
-	cache.insert(key, true)
-	return signature, nil
+	return false
 }
 
-// VerifyThresholdSignatureForMessageSet verifies a threshold signature against a set of message hashes.
-func (cache *cache) VerifyThresholdSignatureForMessageSet(signature consensus.ThresholdSignature, hashes map[hotstuff.ID]consensus.Hash) bool {
-	if signature == nil {
-		return false
-	}
-	var key consensus.Hash
-	hash := sha256.New()
-	for _, h := range hashes {
-		hash.Write(h[:])
-	}
-	hash.Write(signature.ToBytes())
-	hash.Sum(key[:0])
-	if cache.check(key, true) {
+// BatchVerify verifies the given quorum signature against the batch of messages.
+func (cache *cache) BatchVerify(signature consensus.QuorumSignature, batch map[hotstuff.ID][]byte) bool {
+	// get a sorted list of the ids in the Messages map
+	ids := make([]hotstuff.ID, 0, len(batch))
+	for id := range batch {
```

Comment:
Can't recall having seen this construction before to sort the map keys, but it looks more efficient than to first loop over the map to create a slice before sorting. That said, I wonder if it would be better to make this a function called `SortedIDs` that takes a map with `hotstuff.ID` and possibly any type (when we use generics -- assuming we need to sort IDs from maps with other types.) 

For the record, there isn't coming a `SortedKeys` in the stdlib yet, but maybe in [the future.](https://github.com/golang/go/issues/53196)

Reply:
I played around with some different ways to make this generic. However, I think that that using the generic `maps.Keys` function from the experimental `maps` package to extract the keys to a slice, and then the generic `slices.Sort` function from the experimental `slices` package to sort the keys is easier and likely faster.

Reply:
Here are some benchmark results:

```
cpu: AMD Ryzen 9 5900HS with Radeon Graphics        
BenchmarkKeysAsc-16      	       4	 282953742 ns/op	 2238214 B/op	     979 allocs/op
BenchmarkSliceSort-16    	     104	  11535986 ns/op	  858424 B/op	      42 allocs/op
```

The first one, `KeysAsc`, uses the sort.Search method in the same way as the code above.
The second one, `SliceSort`, first uses `maps.Keys` and then sorts them with `slices.Sort`.

Both are sorting the keys of a map with 100 000 entries

Reply:
When sorting maps with only 10 entries, the `slices.Sort` version still wins:

```
cpu: AMD Ryzen 9 5900HS with Radeon Graphics        
BenchmarkKeysAsc-16      	 5160612	       230.5 ns/op	      80 B/op	       1 allocs/op
BenchmarkSliceSort-16    	 7871254	       152.8 ns/op	      80 B/op	       1 allocs/op
```